MIDAS/src/finite-state-machines/fsm.cpp

```diff
@@ -0,0 +1,367 @@
+            }
+            break;
+
+        case STATE_MAIN:
+            // if slowed down enough then go on to the next state
+            if (abs(getVerticalSpeed(bar)) <= landed_vertical_speed_threshold) { 
+             	landed_time = current_time;
+             	state.curr_state = STATE_LANDED; 
+            }
+            break;
+
+        case STATE_LANDED:
+            // if the slow speed was too brief then return to previous state
+            if (abs(getVerticalSpeed(bar)) > landed_vertical_speed_threshold && current_time - landed_time > landed_timer_threshold) {
+             	state.curr_state = STATE_MAIN; 
+            }
+            break;
+        
+        default:
+            return state;
+    
+    }
+    return state;
+}
+
+// this is similiar to the previous function but contains less states
+FSMState FSM::tick_fsm_booster(FSMState state, std::array<HighGData, 8> hg, std::array<Barometer, 8> bar){
+
+    double current_time = tick_to_ms(xTaskGetTickCount());
+    
+    switch (state.curr_state)
+    {
+        case STATE_IDLE:
+            if(getAcceleration(hg) > idle_to_first_boost_acceleration_threshold) {
+                launch_time = current_time;
+                state.curr_state = STATE_FIRST_BOOST;
+            }
+            
+            break;
+
+        case STATE_FIRST_BOOST:
+            if(getAcceleration(hg) < idle_to_first_boost_acceleration_threshold && current_time - launch_time < idle_to_first_boost_time_threshold) {
+                 state.curr_state = STATE_IDLE;
+                 break;
+            }
+            if(getAcceleration(hg) < coast_detection_acceleration_threshold) {
+                burnout_time = current_time;
+                state.curr_state  = STATE_BURNOUT;
+            }
+
+            break;
+
+        case STATE_BURNOUT:
+            if(getAcceleration(hg) >= coast_detection_acceleration_threshold && (current_time - burnout_time) < first_boost_to_burnout_time_threshold) { 
+                state.curr_state = STATE_FIRST_BOOST;
+                break;
+            }
+
+            if(current_time - burnout_time > first_boost_to_burnout_time_threshold) { 
+                sustainer_ignition_time = current_time;
+                state.curr_state = STATE_SUSTAINER_IGNITION;
+            }
+            break;
+
+        case STATE_FIRST_SEPERATION: 
+            if (getJerk(hg) < 0) {
+                state.curr_state = STATE_COAST;
+                break;
+            }
+
+            first_seperation_time = current_time;
+
+            if(current_time - first_seperation_time > first_seperation_time_threshold) { 
+	            state.curr_state = STATE_COAST;
+            }
+
+            break;
+            
+        case STATE_COAST:
+            if (getAcceleration(hg) > coast_detection_acceleration_threshold && current_time - coast_time < second_boost_to_coast_time_threshold) {
+                state.curr_state = STATE_SECOND_BOOST;
```

Comment:
STATE_SECOND_BOOST doesnt exist for the booster. This back transition condition also doesn't make sense here. i think that we can remove this entire if statement entirely because if we detect coast too early, the rocket will just either drag separate or hot stage, both of which are fine.